<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Scripts for RNA-seq and ChIP-seq analysis primer • basicBioinformaticsRNI2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Scripts for RNA-seq and ChIP-seq analysis primer">
<meta property="og:description" content="basicBioinformaticsRNI2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basicBioinformaticsRNI2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scripts.html">Scripts for RNA-seq and ChIP-seq analysis primer</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/genomictools">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Scripts for RNA-seq and ChIP-seq analysis primer</h1>
                        <h4 class="author">Jianhong Ou</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/genomictools/blob/master/vignettes/scripts.Rmd"><code>vignettes/scripts.Rmd</code></a></small>
      <div class="hidden name"><code>scripts.Rmd</code></div>

    </div>

    
    
<div id="learning-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning Objectives:</h2>
<ol style="list-style-type: decimal">
<li><p>pipeline: kallisto and Salmon + tximport + DESeq2 for RNA-seq</p></li>
<li><p>pipeline: bwa + MACS2 for ChIP-seq</p></li>
</ol>
</div>
<div id="set-up" class="section level2">
<h2 class="hasAnchor">
<a href="#set-up" class="anchor"></a>Set-up</h2>
<p>All the materials used in this workshop is included in <a href="https://www.docker.com/">Docker</a> file: <a href="https://hub.docker.com/repository/docker/jianhong/genomictools">jianhong/genomictools</a></p>
<p>Before we start, please download Docker and install Docker Desktop.</p>
<div id="install-docker" class="section level3">
<h3 class="hasAnchor">
<a href="#install-docker" class="anchor"></a>Install docker</h3>
<p>Once the docker installed on you system, please try to run the following code in a terminal. This script will get the latest version of docker container for this workshop.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r">which docker
docker pull jianhong/genomictools:latest</pre></body></html></div>
<p>Change the docker Memory Resources to &gt;= 5G in the Preferences setting page of Docker. And then run the following code in a terminal.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r">## set docker memory to &gt; 5G
## set docker memory to &gt; 5G ; !important
cd ~
mkdir tmp4genomictools
docker run -e PASSWORD=123456 -p 8787:8787 \
  -v ${PWD}/tmp4genomictools:/home/rstudio \
  jianhong/genomictools:latest</pre></body></html></div>
<p>“docker run” will open port 8787 on your system. Now you can login rstudio in the container at <a href="http://localhost:8787" class="uri">http://localhost:8787</a> with username “rstudio” and password “123456”. All the following steps are running in Rstudio “Terminal” or “Console”.</p>
</div>
</div>
<div id="run-kallisto-and-salmon-for-rna-seq" class="section level2">
<h2 class="hasAnchor">
<a href="#run-kallisto-and-salmon-for-rna-seq" class="anchor"></a>Run kallisto and Salmon for RNA-seq</h2>
<p>Run the following code Console panel of rstudio. The sample files are packaged in basicBioinformaticsRNI2020 package.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"basicBioinformaticsRNI2020"</span>)</pre></body></html></div>
<p>And then run following code in “Terminal” panel of rstudio. Here the path is the output of above code. It will copy the sample files into current working directory.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r">path="/Library/Frameworks/R.framework/Versions/4.1/Resources/library/basicBioinformaticsRNI2020/extdata"
cp -r $path ./</pre></body></html></div>
<p>Now we will download the zebrafish cDNA files from ENSEMBL in order to build the Kallisto and Salmon transcript index. If you are doing rRNA depletion library, please download and merge the cDNA and ncDNA files from ENSEMBL to make the full transcriptome.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r">cd extdata/RNAseq
wget ftp://ftp.ensembl.org/pub/release-100/fasta/danio_rerio/cdna/Danio_rerio.GRCz11.cdna.all.fa.gz</pre></body></html></div>
<p>Now we can build the transcriptome index. It will take some time and memory. This is the reason why we need to set docker memory to &gt; 5G.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r">kallisto index -i danRer.GRCz11_transcrits.idx Danio_rerio.GRCz11.cdna.all.fa.gz

salmon index -i danRer.GRCz11_transcrits.salmon.idx -t Danio_rerio.GRCz11.cdna.all.fa.gz</pre></body></html></div>
<p>It’s time for quantifying the FASTQ files against our Kallisto index and Salmon index. We run both of them for comparison. For real data, you can select one of them.</p>
<p>Because our FASTQ files are single end reads, we need to set –single for kallisto. And we also need to give the fragment length and standard error of the length for kallisto.</p>
<p>For Salmon, now we can set library type to “A”.</p>
<p>In this example, we put “-t 2” and “-p 2” so we can use up to two processors in the bootstrapping. We set the bootstrapping by “-b 30” and “–numBootstraps 30”.</p>
<p>It will take several minutes for the sample data.</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r">mkdir -p kallisto_quant
mkdir -p salmon_quant
for rep in 1 2
do
for cond in Ablated Uninjured
do
kallisto quant -i danRer.GRCz11_transcrits.idx \
               -o kallisto_quant/$cond.rep$rep \
               -b 30 -t 2 fastq/$cond.rep$rep.fastq.gz \
               --single -l 200 -s 50
salmon quant -i danRer.GRCz11_transcrits.salmon.idx -l A \
             -r fastq/$cond.rep$rep.fastq.gz \
             --validateMappings -p 2 \
             -o salmon_quant/$cond.rep$rep \
             --numBootstraps 30 --seqBias --gcBias
done
done</pre></body></html></div>
</div>
<div id="r-scripts-for-rnaseq" class="section level2">
<h2 class="hasAnchor">
<a href="#r-scripts-for-rnaseq" class="anchor"></a>R scripts for RNAseq</h2>
<p>Here we give an example workflow for a differential expression (DE) analysis. The following code should be run in Console of Rstudio.</p>
<div id="prepare-the-transcripts-to-genes-map-table" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-the-transcripts-to-genes-map-table" class="anchor"></a>prepare the transcripts to genes map table</h3>
<p>We are trying to do gene level DE analysis. We need to aggregate the transcript level counts to gene level. We borrowed from the Sleuth documentation by retrieve ENSEMBL transcript id to gene id.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biomaRt</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="no">tx2gene</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">species</span><span class="kw">=</span><span class="st">"hsapiens"</span>, <span class="no">...</span>){
  <span class="no">mart</span> <span class="kw">&lt;-</span> <span class="kw pkg">biomaRt</span><span class="kw ns">::</span><span class="fu">useMart</span>(<span class="kw">biomart</span> <span class="kw">=</span> <span class="st">"ensembl"</span>,
                           <span class="kw">dataset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">species</span>, <span class="st">"_gene_ensembl"</span>), <span class="no">...</span>)
  <span class="no">t2g</span> <span class="kw">&lt;-</span> <span class="kw pkg">biomaRt</span><span class="kw ns">::</span><span class="fu">getBM</span>(<span class="kw">attributes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ensembl_transcript_id"</span>,
                                       <span class="st">"ensembl_gene_id"</span>,
                                       <span class="st">"external_gene_name"</span>), <span class="kw">mart</span> <span class="kw">=</span> <span class="no">mart</span>)
  <span class="no">t2g</span> <span class="kw">&lt;-</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu">rename</span>(<span class="no">t2g</span>, <span class="kw">target_id</span> <span class="kw">=</span> <span class="no">ensembl_transcript_id</span>,
                       <span class="kw">ens_gene</span> <span class="kw">=</span> <span class="no">ensembl_gene_id</span>, <span class="kw">ext_gene</span> <span class="kw">=</span> <span class="no">external_gene_name</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">t2g</span>)
}

<span class="no">t2g</span> <span class="kw">&lt;-</span> <span class="fu">tx2gene</span>(<span class="st">"drerio"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">t2g</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">3</span>)</pre></body></html></div>
<p>The following codes are using tximport to import the transcripts counts and aggregate to gene level. Downstream are using DESeq2 to do gene level DE analysis.</p>
</div>
<div id="run-tximport-deseq2-for-salmon-results" class="section level3">
<h3 class="hasAnchor">
<a href="#run-tximport-deseq2-for-salmon-results" class="anchor"></a>run tximport + DESeq2 for Salmon results</h3>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tximport</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DESeq2</span>)

<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">"boostcamp/RNAseq/"</span>)
(<span class="no">salmon_files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"salmon_quant"</span>, <span class="st">"sf$"</span>, <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>))

<span class="no">txi.salmon</span> <span class="kw">&lt;-</span> <span class="fu">tximport</span>(<span class="no">salmon_files</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"salmon"</span>, <span class="kw">tx2gene</span> <span class="kw">=</span> <span class="no">t2g</span>, <span class="kw">ignoreTxVersion</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="no">sampleTable</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">condition</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">".rep."</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span>(<span class="no">salmon_files</span>))))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sampleTable</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">txi.salmon</span>$<span class="no">counts</span>)
<span class="no">dds.salmon</span> <span class="kw">&lt;-</span> <span class="fu">DESeqDataSetFromTximport</span>(<span class="no">txi.salmon</span>, <span class="no">sampleTable</span>, ~<span class="no">condition</span>)
<span class="no">dds.salmon</span> <span class="kw">&lt;-</span> <span class="fu">DESeq</span>(<span class="no">dds.salmon</span>)
<span class="no">res.salmon</span> <span class="kw">&lt;-</span> <span class="fu">results</span>(<span class="no">dds.salmon</span>)
<span class="no">res.salmon</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">res.salmon</span>$<span class="no">padj</span>), ]</pre></body></html></div>
</div>
<div id="run-tximport-deseq2-for-kallisto-results" class="section level3">
<h3 class="hasAnchor">
<a href="#run-tximport-deseq2-for-kallisto-results" class="anchor"></a>run tximport + DESeq2 for kallisto results</h3>
<div class="sourceCode" id="cb10"><html><body><pre class="r">(<span class="no">kallisto_files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"kallisto_quant"</span>, <span class="st">"abundance.h5"</span>, <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>))

<span class="no">txi.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">tximport</span>(<span class="no">kallisto_files</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"kallisto"</span>, <span class="kw">tx2gene</span> <span class="kw">=</span> <span class="no">t2g</span>, <span class="kw">ignoreTxVersion</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="no">sampleTable</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">condition</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">".rep."</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span>(<span class="no">kallisto_files</span>))))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sampleTable</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">txi.kallisto</span>$<span class="no">counts</span>)
<span class="no">dds.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">DESeqDataSetFromTximport</span>(<span class="no">txi.kallisto</span>, <span class="no">sampleTable</span>, ~<span class="no">condition</span>)
<span class="no">dds.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">DESeq</span>(<span class="no">dds.kallisto</span>)
<span class="no">res.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">results</span>(<span class="no">dds.kallisto</span>)
<span class="no">res.kallisto</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">res.kallisto</span>$<span class="no">padj</span>), ]</pre></body></html></div>
</div>
<div id="run-sleuth" class="section level3">
<h3 class="hasAnchor">
<a href="#run-sleuth" class="anchor"></a>run Sleuth</h3>
<p>The following code is showing how to use sleuth to do DE analysis.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="co">#BiocManager::install("pachterlab/sleuth")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sleuth</span>)
<span class="no">samples</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"kallisto_quant"</span>)
<span class="no">s2c</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">path</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"kallisto_quant"</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                  <span class="kw">sample</span><span class="kw">=</span><span class="no">samples</span>,
                  <span class="kw">condition</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">".rep."</span>, <span class="st">""</span>, <span class="no">samples</span>))
<span class="no">s2c</span>
<span class="no">so</span> <span class="kw">&lt;-</span> <span class="fu">sleuth_prep</span>(<span class="no">s2c</span>, ~-<span class="fl">1</span>+<span class="no">condition</span>, <span class="kw">target_mapping</span> <span class="kw">=</span> <span class="no">t2g</span>)
<span class="no">so</span> <span class="kw">&lt;-</span> <span class="fu">sleuth_fit</span>(<span class="no">so</span>)
<span class="co">## which_beta must be in the colname of design table.</span>
<span class="no">so</span> <span class="kw">&lt;-</span> <span class="fu">sleuth_wt</span>(<span class="no">so</span>, <span class="kw">which_beta</span> <span class="kw">=</span> <span class="st">"conditionAblated"</span>)
<span class="fu">sleuth_live</span>(<span class="no">so</span>)</pre></body></html></div>
</div>
</div>
<div id="scripts-for-chipseq" class="section level2">
<h2 class="hasAnchor">
<a href="#scripts-for-chipseq" class="anchor"></a>scripts for ChIPseq</h2>
<div id="prepare-index-file-for-bwa" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-index-file-for-bwa" class="anchor"></a>prepare index file for bwa</h3>
<p>Run following code in Terminal panel of rstudio.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r">wget ftp://ftp.ensembl.org/pub/release-100/fasta/danio_rerio/dna/Danio_rerio.GRCz11.dna.primary_assembly.fa.gz
bwa index -p GRCz11 Danio_rerio.GRCz11.dna.primary_assembly.fa.gz</pre></body></html></div>
</div>
<div id="run-fastqc-mapping-and-macs2" class="section level3">
<h3 class="hasAnchor">
<a href="#run-fastqc-mapping-and-macs2" class="anchor"></a>Run fastQC, mapping and MACS2</h3>
<div class="sourceCode" id="cb13"><html><body><pre class="r">group=(Ablated Ablated Uninjured Uninjured)
tag=(Ablated.rep1 Ablated.rep2 Uninjured.rep1 Uninjured.rep2)
species=danRer11
prefix=bwa

for i in {0..3}
do
## fastQC
mkdir -p fastqc/${group[$i]}
fastqc -o fastqc/${group[$i]} -t 2 \
       fastq/${tag[$i]}.fastq.gz
## trim adapter, need trim_galore be installed, here we do not do this step
# mkdir -p fastq.trimmed
# trim_galore -q 15 --fastqc -o fastq.trimmed/${group[$i]} fastq/${tag[$i]}.fastq.gz

## mapping by bwa
mkdir -p sam
bwa mem -M GRCz11 -t 2 \
           fastq/${tag[$i]}.fastq.gz \
           &gt; sam/$prefix.$species.${tag[$i]}.sam \
           2&gt; bwa.$prefix.$species.${tag[$i]}.log.txt

mkdir -p bam
samtools view -bhS -q 30 sam/$prefix.$species.${tag[$i]}.sam &gt; bam/$prefix.$species.${tag[$i]}.bam
samtools sort bam/$prefix.$species.${tag[$i]}.bam -o bam/$prefix.$species.${tag[$i]}.srt.bam
samtools index bam/$prefix.$species.${tag[$i]}.srt.bam
rm bam/$prefix.$species.${tag[$i]}.bam

mkdir -p bw
bamCoverage -b bam/$prefix.$species.${tag[$i]}.srt.bam -o bw/$prefix.$species.${tag[$i]}.bw --normalizeUsing CPM


mkdir -p bam/picard
picard MarkDuplicates \
       INPUT=bam/$prefix.$species.${tag[$i]}.srt.bam \
       OUTPUT=bam/$prefix.$species.${tag[$i]}.srt.markDup.bam \
       METRICS_FILE=bam/picard/$prefix.$species.${tag[$i]}.srt.fil.picard_info.txt \
       REMOVE_DUPLICATES=true ASSUME_SORTED=true VALIDATION_STRINGENCY=LENIENT
samtools index bam/$prefix.$species.${tag[$i]}.srt.markDup.bam


## call peaks
mkdir -p macs2/${tag[$i]}

macs2 callpeak -t bam/${prefix}.$species.${tag[$i]}.srt.markDup.bam \
               -f BAM -g 1.2e9 -n ${prefix}.$species.${tag[$i]} \
               --outdir macs2/${tag[$i]} -q 0.05 \
               -B --nomodel --extsize 150

done</pre></body></html></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
