<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Scripts for RNA-seq and ChIP-seq analysis primer • basicBioinformaticsRNI2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Scripts for RNA-seq and ChIP-seq analysis primer">
<meta property="og:description" content="basicBioinformaticsRNI2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">basicBioinformaticsRNI2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.0.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/scripts.html">Scripts for RNA-seq and ChIP-seq analysis primer</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jianhong/genomictools">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Scripts for RNA-seq and ChIP-seq analysis primer</h1>
                        <h4 class="author">Jianhong Ou</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jianhong/genomictools/blob/master/vignettes/scripts.Rmd"><code>vignettes/scripts.Rmd</code></a></small>
      <div class="hidden name"><code>scripts.Rmd</code></div>

    </div>

    
    
<div id="learning-objectives" class="section level2">
<h2 class="hasAnchor">
<a href="#learning-objectives" class="anchor"></a>Learning Objectives:</h2>
<ol style="list-style-type: decimal">
<li><p>pipeline: kallisto and Salmon + tximport + DESeq2 for RNA-seq</p></li>
<li><p>pipeline: bwa + MACS2 for ChIP-seq</p></li>
</ol>
<p>For more information please refer to the <a href="https://github.com/jianhong/genomictools/blob/master/inst/extdata/BasicBioinformaticsRNAseqChIPseqPrimer.pdf">slides</a></p>
</div>
<div id="set-up" class="section level2">
<h2 class="hasAnchor">
<a href="#set-up" class="anchor"></a>Set-up</h2>
<p>All the materials used in this workshop is included in <a href="https://www.docker.com/">Docker</a> file: <a href="https://hub.docker.com/repository/docker/jianhong/genomictools">jianhong/genomictools</a></p>
<p>Before we start, please download Docker and install Docker Desktop.</p>
<div id="install-docker" class="section level3">
<h3 class="hasAnchor">
<a href="#install-docker" class="anchor"></a>Install docker</h3>
<p>Once the docker installed on you system, please try to run the following code in a terminal. This script will get the latest version of docker container for this workshop.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="fu">which</span> docker</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="ex">docker</span> pull jianhong/genomictools:latest</a></code></pre></div>
<p>Change the docker Memory Resources to &gt;= 5G in the Preferences setting page of Docker. And then run the following code in a terminal.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co">## set docker memory to &gt; 5G</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="co">## set docker memory to &gt; 5G ; !important</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="bu">cd</span> ~</a>
<a class="sourceLine" id="cb2-4" data-line-number="4"><span class="fu">mkdir</span> tmp4genomictools</a>
<a class="sourceLine" id="cb2-5" data-line-number="5"><span class="ex">docker</span> run -e PASSWORD=123456 -p 8787:8787 \</a>
<a class="sourceLine" id="cb2-6" data-line-number="6">  -v <span class="va">${PWD}</span>/tmp4genomictools:/home/rstudio \</a>
<a class="sourceLine" id="cb2-7" data-line-number="7">  jianhong/genomictools:latest</a></code></pre></div>
<p>“docker run” will open port 8787 on your system. Now you can login rstudio in the container at <a href="http://localhost:8787" class="uri">http://localhost:8787</a> with username “rstudio” and password “123456”. All the following steps are running in Rstudio “Terminal” or “Console”.</p>
</div>
</div>
<div id="run-kallisto-and-salmon-for-rna-seq" class="section level2">
<h2 class="hasAnchor">
<a href="#run-kallisto-and-salmon-for-rna-seq" class="anchor"></a>Run kallisto and Salmon for RNA-seq</h2>
<p>The sample files are packaged in basicBioinformaticsRNI2020 package and Docker container.</p>
<p>Now we will download the zebrafish cDNA files from ENSEMBL in order to build the Kallisto and Salmon transcript index. If you are doing rRNA depletion library, please download and merge the cDNA and ncDNA files from ENSEMBL to make the full transcriptome.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="bu">cd</span> RNAseq</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="fu">wget</span> ftp://ftp.ensembl.org/pub/release-100/fasta/danio_rerio/cdna/Danio_rerio.GRCz11.cdna.all.fa.gz</a></code></pre></div>
<p>Now we can build the transcriptome index. It will take some time and memory. This is the reason why we need to set docker memory to &gt; 5G.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="ex">kallisto</span> index -i danRer.GRCz11_transcrits.idx Danio_rerio.GRCz11.cdna.all.fa.gz</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3"><span class="ex">salmon</span> index -i danRer.GRCz11_transcrits.salmon.idx -t Danio_rerio.GRCz11.cdna.all.fa.gz</a></code></pre></div>
<p>It’s time for quantifying the FASTQ files against our Kallisto index and Salmon index. We run both of them for comparison. For real data, you can select one of them.</p>
<p>Because our FASTQ files are single end reads, we need to set –single for kallisto. And we also need to give the fragment length and standard error of the length for kallisto.</p>
<p>For Salmon, now we can set library type to “A”.</p>
<p>In this example, we put “-t 2” and “-p 2” so we can use up to two processors in the bootstrapping. We set the bootstrapping by “-b 30” and “–numBootstraps 30”.</p>
<p>It will take several minutes for the sample data.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="fu">mkdir</span> -p kallisto_quant</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="fu">mkdir</span> -p salmon_quant</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw">for</span> <span class="ex">rep</span> in 1 2</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="kw">do</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">for</span> <span class="ex">cond</span> in Ablated Uninjured</a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="kw">do</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="ex">kallisto</span> quant -i danRer.GRCz11_transcrits.idx \</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">               -o kallisto_quant/<span class="va">$cond</span>.rep<span class="va">$rep</span> \</a>
<a class="sourceLine" id="cb5-9" data-line-number="9">               -b 30 -t 2 fastq/<span class="va">$cond</span>.rep<span class="va">$rep</span>.fastq.gz \</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">               --single -l 200 -s 50</a>
<a class="sourceLine" id="cb5-11" data-line-number="11"><span class="ex">salmon</span> quant -i danRer.GRCz11_transcrits.salmon.idx -l A \</a>
<a class="sourceLine" id="cb5-12" data-line-number="12">             -r fastq/<span class="va">$cond</span>.rep<span class="va">$rep</span>.fastq.gz \</a>
<a class="sourceLine" id="cb5-13" data-line-number="13">             --validateMappings -p 2 \</a>
<a class="sourceLine" id="cb5-14" data-line-number="14">             -o salmon_quant/<span class="va">$cond</span>.rep<span class="va">$rep</span> \</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">             --numBootstraps 30 --seqBias --gcBias</a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="kw">done</span></a>
<a class="sourceLine" id="cb5-17" data-line-number="17"><span class="kw">done</span></a></code></pre></div>
</div>
<div id="r-scripts-for-rnaseq" class="section level2">
<h2 class="hasAnchor">
<a href="#r-scripts-for-rnaseq" class="anchor"></a>R scripts for RNAseq</h2>
<p>Here we give an example workflow for a differential expression (DE) analysis. The following code should be run in Console of Rstudio.</p>
<div id="prepare-the-transcripts-to-genes-map-table" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-the-transcripts-to-genes-map-table" class="anchor"></a>prepare the transcripts to genes map table</h3>
<p>We are trying to do gene level DE analysis. We need to aggregate the transcript level counts to gene level. We borrowed from the Sleuth documentation by retrieve ENSEMBL transcript id to gene id.</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">biomaRt</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">dplyr</span>)
<span class="no">tx2gene</span> <span class="kw">&lt;-</span> <span class="kw">function</span>(<span class="no">species</span><span class="kw">=</span><span class="st">"hsapiens"</span>, <span class="no">...</span>){
  <span class="no">mart</span> <span class="kw">&lt;-</span> <span class="kw pkg">biomaRt</span><span class="kw ns">::</span><span class="fu">useMart</span>(<span class="kw">biomart</span> <span class="kw">=</span> <span class="st">"ensembl"</span>,
                           <span class="kw">dataset</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="no">species</span>, <span class="st">"_gene_ensembl"</span>), <span class="no">...</span>)
  <span class="no">t2g</span> <span class="kw">&lt;-</span> <span class="kw pkg">biomaRt</span><span class="kw ns">::</span><span class="fu">getBM</span>(<span class="kw">attributes</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"ensembl_transcript_id"</span>,
                                       <span class="st">"ensembl_gene_id"</span>,
                                       <span class="st">"external_gene_name"</span>), <span class="kw">mart</span> <span class="kw">=</span> <span class="no">mart</span>)
  <span class="no">t2g</span> <span class="kw">&lt;-</span> <span class="kw pkg">dplyr</span><span class="kw ns">::</span><span class="fu">rename</span>(<span class="no">t2g</span>, <span class="kw">target_id</span> <span class="kw">=</span> <span class="no">ensembl_transcript_id</span>,
                       <span class="kw">ens_gene</span> <span class="kw">=</span> <span class="no">ensembl_gene_id</span>, <span class="kw">ext_gene</span> <span class="kw">=</span> <span class="no">external_gene_name</span>)
  <span class="fu"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="no">t2g</span>)
}

<span class="no">t2g</span> <span class="kw">&lt;-</span> <span class="fu">tx2gene</span>(<span class="st">"drerio"</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="no">t2g</span>, <span class="kw">n</span><span class="kw">=</span><span class="fl">3</span>)</pre></body></html></div>
<p>The following codes are using tximport to import the transcripts counts and aggregate to gene level. Downstream are using DESeq2 to do gene level DE analysis.</p>
</div>
<div id="run-tximport-deseq2-for-salmon-results" class="section level3">
<h3 class="hasAnchor">
<a href="#run-tximport-deseq2-for-salmon-results" class="anchor"></a>run tximport + DESeq2 for Salmon results</h3>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tximport</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">DESeq2</span>)

<span class="fu"><a href="https://rdrr.io/r/base/getwd.html">setwd</a></span>(<span class="st">"boostcamp/RNAseq/"</span>)
(<span class="no">salmon_files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"salmon_quant"</span>, <span class="st">"sf$"</span>,
                     <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>,
                     <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>))

<span class="no">txi.salmon</span> <span class="kw">&lt;-</span> <span class="fu">tximport</span>(<span class="no">salmon_files</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"salmon"</span>,
                       <span class="kw">tx2gene</span> <span class="kw">=</span> <span class="no">t2g</span>, <span class="kw">ignoreTxVersion</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="no">sampleTable</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">condition</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">".rep."</span>, <span class="st">""</span>,
                                          <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span>(<span class="no">salmon_files</span>))))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sampleTable</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">txi.salmon</span>$<span class="no">counts</span>)
<span class="no">dds.salmon</span> <span class="kw">&lt;-</span> <span class="fu">DESeqDataSetFromTximport</span>(<span class="no">txi.salmon</span>, <span class="no">sampleTable</span>, ~<span class="no">condition</span>)
<span class="no">dds.salmon</span> <span class="kw">&lt;-</span> <span class="fu">DESeq</span>(<span class="no">dds.salmon</span>)
<span class="no">res.salmon</span> <span class="kw">&lt;-</span> <span class="fu">results</span>(<span class="no">dds.salmon</span>)
<span class="no">res.salmon</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">res.salmon</span>$<span class="no">padj</span>), ]</pre></body></html></div>
</div>
<div id="run-tximport-deseq2-for-kallisto-results" class="section level3">
<h3 class="hasAnchor">
<a href="#run-tximport-deseq2-for-kallisto-results" class="anchor"></a>run tximport + DESeq2 for kallisto results</h3>
<div class="sourceCode" id="cb8"><html><body><pre class="r">(<span class="no">kallisto_files</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"kallisto_quant"</span>, <span class="st">"abundance.h5"</span>,
                       <span class="kw">recursive</span> <span class="kw">=</span> <span class="fl">TRUE</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>))

<span class="no">txi.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">tximport</span>(<span class="no">kallisto_files</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"kallisto"</span>,
                         <span class="kw">tx2gene</span> <span class="kw">=</span> <span class="no">t2g</span>, <span class="kw">ignoreTxVersion</span><span class="kw">=</span><span class="fl">TRUE</span>)

<span class="no">sampleTable</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">condition</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">".rep."</span>, <span class="st">""</span>,
                                          <span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/basename.html">dirname</a></span>(<span class="no">kallisto_files</span>))))
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span>(<span class="no">sampleTable</span>) <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(<span class="no">txi.kallisto</span>$<span class="no">counts</span>)
<span class="no">dds.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">DESeqDataSetFromTximport</span>(<span class="no">txi.kallisto</span>,
                                         <span class="no">sampleTable</span>, ~<span class="no">condition</span>)
<span class="no">dds.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">DESeq</span>(<span class="no">dds.kallisto</span>)
<span class="no">res.kallisto</span> <span class="kw">&lt;-</span> <span class="fu">results</span>(<span class="no">dds.kallisto</span>)
<span class="no">res.kallisto</span>[!<span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">res.kallisto</span>$<span class="no">padj</span>), ]</pre></body></html></div>
</div>
<div id="run-sleuth" class="section level3">
<h3 class="hasAnchor">
<a href="#run-sleuth" class="anchor"></a>run Sleuth</h3>
<p>The following code is showing how to use sleuth to do DE analysis.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="co">#BiocManager::install("pachterlab/sleuth")</span>
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">sleuth</span>)
<span class="no">samples</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"kallisto_quant"</span>)
<span class="no">s2c</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="kw">path</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html">dir</a></span>(<span class="st">"kallisto_quant"</span>, <span class="kw">full.names</span> <span class="kw">=</span> <span class="fl">TRUE</span>),
                  <span class="kw">sample</span><span class="kw">=</span><span class="no">samples</span>,
                  <span class="kw">condition</span><span class="kw">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">sub</a></span>(<span class="st">".rep."</span>, <span class="st">""</span>, <span class="no">samples</span>))
<span class="no">s2c</span>
<span class="no">so</span> <span class="kw">&lt;-</span> <span class="fu">sleuth_prep</span>(<span class="no">s2c</span>, ~-<span class="fl">1</span>+<span class="no">condition</span>, <span class="kw">target_mapping</span> <span class="kw">=</span> <span class="no">t2g</span>)
<span class="no">so</span> <span class="kw">&lt;-</span> <span class="fu">sleuth_fit</span>(<span class="no">so</span>)
<span class="co">## which_beta must be in the colname of design table.</span>
<span class="no">so</span> <span class="kw">&lt;-</span> <span class="fu">sleuth_wt</span>(<span class="no">so</span>, <span class="kw">which_beta</span> <span class="kw">=</span> <span class="st">"conditionAblated"</span>)
<span class="fu">sleuth_live</span>(<span class="no">so</span>)</pre></body></html></div>
</div>
</div>
<div id="scripts-for-chipseq" class="section level2">
<h2 class="hasAnchor">
<a href="#scripts-for-chipseq" class="anchor"></a>scripts for ChIPseq</h2>
<div id="prepare-index-file-for-bwa" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-index-file-for-bwa" class="anchor"></a>prepare index file for bwa</h3>
<p>Run following code in Terminal panel of rstudio.</p>
<p>The sample data is located at “/home/rstudio/ChIPseq” folder. It is zebrafish H3 ChIP-seq data.</p>
<p>We will map the reads by BWA (Burrows-Wheeler Aligner). Compare to Bowtie2, BWA is a little slower but a bit more accurate and provides information on which alignments are trustworthy. The bwa-mem mode is generally recommended for high-quality queries. It is not limited by sequence reads size as bwa-backtrack and bwa-sw. Aligning reads with bwa-mem, there are two steps, build the index and do alignment. BWA indexes the genome with an FM index based on the Burrows-Wheeler Transform to keep memory requirements low for the alignment process.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="co">## change you working directory</span></a>
<a class="sourceLine" id="cb10-2" data-line-number="2"><span class="bu">cd</span> /home/rstudio/ChIPseq</a>
<a class="sourceLine" id="cb10-3" data-line-number="3"><span class="co">## download the zebrafish GRCz11 genome from ENSEMBLE</span></a>
<a class="sourceLine" id="cb10-4" data-line-number="4"><span class="fu">wget</span> ftp://ftp.ensembl.org/pub/release-100/fasta/danio_rerio/dna/Danio_rerio.GRCz11.dna.primary_assembly.fa.gz</a>
<a class="sourceLine" id="cb10-5" data-line-number="5"><span class="co">## build the index</span></a>
<a class="sourceLine" id="cb10-6" data-line-number="6"><span class="co">## -p: prefix for all index files</span></a>
<a class="sourceLine" id="cb10-7" data-line-number="7"><span class="ex">bwa</span> index -p GRCz11 Danio_rerio.GRCz11.dna.primary_assembly.fa.gz</a></code></pre></div>
</div>
<div id="run-fastqc-mapping-and-macs2" class="section level3">
<h3 class="hasAnchor">
<a href="#run-fastqc-mapping-and-macs2" class="anchor"></a>Run fastQC, mapping and MACS2</h3>
<p>We will the scripts in a bash loop. The samples are from two uninjured and two ablated fish heart tissue. We use fastQC to check the quality of raw reads. The results will saved to “fastqc” folder. You can go through the results by click the index.html file in each sub-folder. We will perform alignment on the single-end reads. For more information on BWA and its functionality please refer to the <a href="http://bio-bwa.sourceforge.net/bwa.shtml">user manual</a>. <a href="https://github.com/macs3-project/MACS">MACS2</a> will be used to call the peaks.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><a class="sourceLine" id="cb11-1" data-line-number="1"><span class="va">group=(</span>Ablated Ablated Uninjured Uninjured<span class="va">)</span></a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="va">tag=(</span>Ablated.rep1 Ablated.rep2 Uninjured.rep1 Uninjured.rep2<span class="va">)</span></a>
<a class="sourceLine" id="cb11-3" data-line-number="3"><span class="va">species=</span>danRer11</a>
<a class="sourceLine" id="cb11-4" data-line-number="4"><span class="va">prefix=</span>bwa</a>
<a class="sourceLine" id="cb11-5" data-line-number="5"></a>
<a class="sourceLine" id="cb11-6" data-line-number="6"><span class="kw">for</span> <span class="ex">i</span> in <span class="dt">{0..3}</span></a>
<a class="sourceLine" id="cb11-7" data-line-number="7"><span class="kw">do</span></a>
<a class="sourceLine" id="cb11-8" data-line-number="8"><span class="co">## fastQC</span></a>
<a class="sourceLine" id="cb11-9" data-line-number="9"><span class="fu">mkdir</span> -p fastqc/<span class="va">${group[$i]}</span></a>
<a class="sourceLine" id="cb11-10" data-line-number="10"><span class="ex">fastqc</span> -o fastqc/<span class="va">${group[$i]}</span> -t 2 \</a>
<a class="sourceLine" id="cb11-11" data-line-number="11">       fastq/<span class="va">${tag[$i]}</span>.fastq.gz</a>
<a class="sourceLine" id="cb11-12" data-line-number="12"><span class="co">## trim adapter, need trim_galore be installed, here we do not do this step</span></a>
<a class="sourceLine" id="cb11-13" data-line-number="13"><span class="co"># mkdir -p fastq.trimmed</span></a>
<a class="sourceLine" id="cb11-14" data-line-number="14"><span class="co"># trim_galore -q 15 --fastqc -o fastq.trimmed/${group[$i]} fastq/${tag[$i]}.fastq.gz</span></a>
<a class="sourceLine" id="cb11-15" data-line-number="15"></a>
<a class="sourceLine" id="cb11-16" data-line-number="16"><span class="co">## mapping by bwa</span></a>
<a class="sourceLine" id="cb11-17" data-line-number="17"><span class="fu">mkdir</span> -p sam</a>
<a class="sourceLine" id="cb11-18" data-line-number="18"><span class="co">## -t: number of threads</span></a>
<a class="sourceLine" id="cb11-19" data-line-number="19"><span class="co">## -M: mark shorter split hits as secondary, this is optional for Picard compatibility.</span></a>
<a class="sourceLine" id="cb11-20" data-line-number="20"><span class="co">## &gt;: save alignment to a SAM file</span></a>
<a class="sourceLine" id="cb11-21" data-line-number="21"><span class="co">## 2&gt;: save standard error to log file</span></a>
<a class="sourceLine" id="cb11-22" data-line-number="22"><span class="ex">bwa</span> mem -M -t 2 GRCz11 \</a>
<a class="sourceLine" id="cb11-23" data-line-number="23">           fastq/<span class="va">${tag[$i]}</span>.fastq.gz \</a>
<a class="sourceLine" id="cb11-24" data-line-number="24">           <span class="op">&gt;</span> sam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.sam \</a>
<a class="sourceLine" id="cb11-25" data-line-number="25">           <span class="op">2&gt;</span> bwa.<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.log.txt</a>
<a class="sourceLine" id="cb11-26" data-line-number="26"></a>
<a class="sourceLine" id="cb11-27" data-line-number="27"><span class="co">## convert sam file to bam and clean-up</span></a>
<a class="sourceLine" id="cb11-28" data-line-number="28"><span class="fu">mkdir</span> -p bam</a>
<a class="sourceLine" id="cb11-29" data-line-number="29"><span class="co">## -q: skip alignments with MAPQ samller than 30.</span></a>
<a class="sourceLine" id="cb11-30" data-line-number="30"><span class="ex">samtools</span> view -bhS -q 30 sam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.sam <span class="op">&gt;</span> bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.bam</a>
<a class="sourceLine" id="cb11-31" data-line-number="31"><span class="co">## sort and index the bam file for quick access.</span></a>
<a class="sourceLine" id="cb11-32" data-line-number="32"><span class="ex">samtools</span> sort bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.bam -o bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.bam</a>
<a class="sourceLine" id="cb11-33" data-line-number="33"><span class="ex">samtools</span> index bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.bam</a>
<a class="sourceLine" id="cb11-34" data-line-number="34"><span class="co">## remove un-sorted bam file.</span></a>
<a class="sourceLine" id="cb11-35" data-line-number="35"><span class="fu">rm</span> bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.bam</a>
<a class="sourceLine" id="cb11-36" data-line-number="36"></a>
<a class="sourceLine" id="cb11-37" data-line-number="37"><span class="co">## use deeptools::bamCoverage to generate bigwig files</span></a>
<a class="sourceLine" id="cb11-38" data-line-number="38"><span class="co">## the bw file can be viewed in IGV</span></a>
<a class="sourceLine" id="cb11-39" data-line-number="39"><span class="fu">mkdir</span> -p bw</a>
<a class="sourceLine" id="cb11-40" data-line-number="40"><span class="ex">bamCoverage</span> -b bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.bam -o bw/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.bw --normalizeUsing CPM</a>
<a class="sourceLine" id="cb11-41" data-line-number="41"></a>
<a class="sourceLine" id="cb11-42" data-line-number="42"><span class="co">## we remove the duplicated by picard::MarkDuplicates. </span></a>
<a class="sourceLine" id="cb11-43" data-line-number="43"><span class="fu">mkdir</span> -p bam/picard</a>
<a class="sourceLine" id="cb11-44" data-line-number="44"><span class="ex">picard</span> MarkDuplicates \</a>
<a class="sourceLine" id="cb11-45" data-line-number="45">       INPUT=bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.bam \</a>
<a class="sourceLine" id="cb11-46" data-line-number="46">       OUTPUT=bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.markDup.bam \</a>
<a class="sourceLine" id="cb11-47" data-line-number="47">       METRICS_FILE=bam/picard/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.fil.picard_info.txt \</a>
<a class="sourceLine" id="cb11-48" data-line-number="48">       REMOVE_DUPLICATES=true ASSUME_SORTED=true VALIDATION_STRINGENCY=LENIENT</a>
<a class="sourceLine" id="cb11-49" data-line-number="49"><span class="ex">samtools</span> index bam/<span class="va">$prefix</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.markDup.bam</a>
<a class="sourceLine" id="cb11-50" data-line-number="50"></a>
<a class="sourceLine" id="cb11-51" data-line-number="51"></a>
<a class="sourceLine" id="cb11-52" data-line-number="52"><span class="co">## call peaks by macs2</span></a>
<a class="sourceLine" id="cb11-53" data-line-number="53"><span class="fu">mkdir</span> -p macs2/<span class="va">${tag[$i]}</span></a>
<a class="sourceLine" id="cb11-54" data-line-number="54"><span class="co">## -g: mappable genome size</span></a>
<a class="sourceLine" id="cb11-55" data-line-number="55"><span class="co">## -q: use minimum FDR 0.05 cutoff to call significant regions.</span></a>
<a class="sourceLine" id="cb11-56" data-line-number="56"><span class="co">## -B: ask MACS2 to output bedGraph files for experiment.</span></a>
<a class="sourceLine" id="cb11-57" data-line-number="57"><span class="co">## --nomodel --extsize 150: the subset data is not big enough (&lt;1000 peak) for</span></a>
<a class="sourceLine" id="cb11-58" data-line-number="58"><span class="co">## macs2 to generate a model. We manually feed one.</span></a>
<a class="sourceLine" id="cb11-59" data-line-number="59"><span class="ex">macs2</span> callpeak -t bam/<span class="va">${prefix}</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span>.srt.markDup.bam \</a>
<a class="sourceLine" id="cb11-60" data-line-number="60">               -f BAM -g 1.4e9 -n <span class="va">${prefix}</span>.<span class="va">$species</span>.<span class="va">${tag[$i]}</span> \</a>
<a class="sourceLine" id="cb11-61" data-line-number="61">               --outdir macs2/<span class="va">${tag[$i]}</span> -q 0.05 \</a>
<a class="sourceLine" id="cb11-62" data-line-number="62">               -B --nomodel --extsize 150</a>
<a class="sourceLine" id="cb11-63" data-line-number="63"></a>
<a class="sourceLine" id="cb11-64" data-line-number="64"><span class="kw">done</span></a></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jianhong Ou.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
